File: src/claude_recall/embeddings.py
"""Ollama embedding client + vector ops (v0.3).

Part of the optional ``[embeddings]`` extra. Requires ``numpy`` and ``httpx``.
Core ``claude-recall`` stays zero-dep; importing this module fails cleanly if
the extra isn't installed, letting callers guard with a try/except.

See docs/EMBEDDINGS-PLAN.md for the full design, and docs/PLAN.md §4 + §8 for
the envelope (model, config block, opt-in framing).
"""

from __future__ import annotations

from dataclasses import dataclass

try:
    import httpx
    import numpy as np
except ImportError as exc:  # pragma: no cover - exercised at install time
    raise ImportError(
        "claude-recall embeddings layer requires the [embeddings] extra. "
        "Install with: pip install 'claude-recall[embeddings]'"
    ) from exc

DEFAULT_BASE_URL = "http://localhost:11434"
DEFAULT_MODEL = "nomic-embed-text"
DEFAULT_TIMEOUT = 10.0
DEFAULT_DIM = 768


class EmbeddingError(RuntimeError):
    """Raised for any terminal embedding-path failure.

    Callers that need to degrade gracefully (CLI search, hook path) should
    catch this and fall back to FTS5-only results.
    """


@dataclass
class ProbeResult:
    ollama_reachable: bool
    version: str | None
    model_present: bool
    embed_ok: bool
    dim: int | None
    error: str | None = None


class OllamaClient:
    """Thin synchronous httpx wrapper around Ollama's /api/embed endpoint.

    One instance per CLI invocation is enough. No connection pooling beyond
    httpx defaults; the embed payload size is small and the call count is
    bounded by ``rerank_pool_size`` for searches or corpus size for
    ``claude-recall embed``.
    """

    def __init__(
        self,
        base_url: str = DEFAULT_BASE_URL,
        model: str = DEFAULT_MODEL,
        timeout: float = DEFAULT_TIMEOUT,
    ) -> None:
        self.base_url = base_url.rstrip("/")
        self.model = model
        self.timeout = timeout
        self._client = httpx.Client(timeout=timeout)

    def close(self) -> None:
        self._client.close()

    def __enter__(self) -> OllamaClient:
        return self

    def __exit__(self, *exc) -> None:
        self.close()

    def embed(self, text: str) -> np.ndarray:
        """Embed a single string. Returns a float32 ndarray of shape (dim,)."""
        matrix = self.embed_batch([text])
        return matrix[0]

    def embed_batch(self, texts: list[str]) -> np.ndarray:
        """Embed a list of strings.

        Returns a float32 ndarray of shape (len(texts), dim). Preserves input
        order. Raises EmbeddingError on any HTTP or shape failure; validates
        that every returned vector is finite and has non-zero norm.
        """
        if not texts:
            return np.zeros((0, DEFAULT_DIM), dtype=np.float32)
        try:
            resp = self._client.post(
                f"{self.base_url}/api/embed",
                json={"model": self.model, "input": texts},
            )
            resp.raise_for_status()
            data = resp.json()
        except httpx.HTTPError as exc:
            raise EmbeddingError(f"Ollama request failed: {exc}") from exc
        except ValueError as exc:
            raise EmbeddingError(f"Ollama returned non-JSON: {exc}") from exc

        raw = data.get("embeddings")
        if not isinstance(raw, list) or len(raw) != len(texts):
            raise EmbeddingError(
                f"Ollama returned {len(raw) if isinstance(raw, list) else '?'} "
                f"embeddings for {len(texts)} inputs"
            )

        matrix = np.asarray(raw, dtype=np.float32)
        if matrix.ndim != 2:
            raise EmbeddingError(
                f"Unexpected embedding shape: {matrix.shape}"
            )
        if not np.isfinite(matrix).all():
            raise EmbeddingError("Ollama returned non-finite values")
        norms = np.linalg.norm(matrix, axis=1)
        if (norms == 0).any():
            raise EmbeddingError("Ollama returned zero-norm vector(s)")
        return matrix

    def probe(self) -> ProbeResult:
        """Diagnose the full embedding path without raising.

        Returns a ProbeResult with per-step booleans so the CLI can tell the
        user exactly which step failed. Never raises.
        """
        version: str | None = None
        try:
            v = self._client.get(f"{self.base_url}/api/version")
            v.raise_for_status()
            version = v.json().get("version")
        except httpx.HTTPError as exc:
            return ProbeResult(
                ollama_reachable=False,
                version=None,
                model_present=False,
                embed_ok=False,
                dim=None,
                error=f"cannot reach Ollama at {self.base_url}: {exc}",
            )

        try:
            t = self._client.get(f"{self.base_url}/api/tags")
            t.raise_for_status()
            tags = t.json().get("models", [])
        except httpx.HTTPError as exc:
            return ProbeResult(
                ollama_reachable=True,
                version=version,
                model_present=False,
                embed_ok=False,
                dim=None,
                error=f"cannot list Ollama models: {exc}",
            )
        names = {m.get("name", "").split(":")[0] for m in tags}
        model_name = self.model.split(":")[0]
        if model_name not in names:
            return ProbeResult(
                ollama_reachable=True,
                version=version,
                model_present=False,
                embed_ok=False,
                dim=None,
                error=(
                    f"model {self.model!r} not pulled. "
                    f"Run: ollama pull {self.model}"
                ),
            )

        try:
            vec = self.embed("probe")
        except EmbeddingError as exc:
            return ProbeResult(
                ollama_reachable=True,
                version=version,
                model_present=True,
                embed_ok=False,
                dim=None,
                error=f"embed probe failed: {exc}",
            )

        return ProbeResult(
            ollama_reachable=True,
            version=version,
            model_present=True,
            embed_ok=True,
            dim=int(vec.shape[0]),
            error=None,
        )


# --- vector BLOB + cosine ---------------------------------------------------

def pack_vector(v: np.ndarray) -> bytes:
    """Serialize a float vector to little-endian float32 bytes for SQLite BLOB.

    Deterministic: identical vectors produce identical bytes across platforms.
    """
    arr = np.asarray(v, dtype="<f4")
    if arr.ndim != 1:
        raise ValueError(f"pack_vector expects a 1-D array, got {arr.shape}")
    return arr.tobytes()


def unpack_vector(blob: bytes, dim: int) -> np.ndarray:
    """Deserialize a BLOB produced by pack_vector back into a float32 array."""
    expected = dim * 4
    if len(blob) != expected:
        raise ValueError(
            f"blob length {len(blob)} does not match expected {expected} "
            f"for dim={dim}"
        )
    return np.frombuffer(blob, dtype="<f4").copy()  # copy so it's writable


def cosine_matrix(query: np.ndarray, candidates: np.ndarray) -> np.ndarray:
    """Cosine similarity between a single query and a matrix of candidates.

    ``query``: shape (D,). ``candidates``: shape (N, D). Returns shape (N,)
    with values in roughly [-1, 1]. Handles zero-norm defensively (rare after
    OllamaClient validation, but cheap).
    """
    if query.ndim != 1:
        raise ValueError(f"query must be 1-D, got {query.shape}")
    if candidates.ndim != 2 or candidates.shape[1] != query.shape[0]:
        raise ValueError(
            f"candidates shape {candidates.shape} incompatible with "
            f"query shape {query.shape}"
        )
    qn = float(np.linalg.norm(query))
    if qn == 0.0:
        return np.zeros(candidates.shape[0], dtype=np.float32)
    cn = np.linalg.norm(candidates, axis=1)
    safe_cn = np.where(cn == 0.0, 1.0, cn)
    dots = candidates @ query
    return (dots / (safe_cn * qn)).astype(np.float32)

File: src/ClaudeRecall.Hook/Embeddings.cs
// Embeddings.cs — port of src/claude_recall/embeddings.py.
//
// Ollama HTTP client (via HttpClient), little-endian float32 BLOB codec
// matching Python's pack_vector, and vectorized cosine similarity over a
// candidate matrix.

using System.Buffers.Binary;
using System.Net.Http.Json;
using System.Numerics;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace ClaudeRecall.Hook;

internal sealed class EmbeddingException : Exception
{
    public EmbeddingException(string message) : base(message) { }
    public EmbeddingException(string message, Exception inner) : base(message, inner) { }
}

internal sealed record ProbeResult(
    bool OllamaReachable,
    string? Version,
    bool ModelPresent,
    bool EmbedOk,
    int? Dim,
    string? Error
);

internal class OllamaClient : IDisposable
{
    public const int DefaultDim = 768;

    private readonly HttpClient _http;
    private readonly string _baseUrl;
    private readonly string _model;
    private readonly string? _keepAlive;

    public OllamaClient(
        string baseUrl,
        string model,
        double timeoutSeconds,
        string? keepAlive = null)
    {
        _baseUrl = baseUrl.TrimEnd('/');
        _model = model;
        _keepAlive = keepAlive;  // issue #11 — sent on every embed call
        _http = new HttpClient
        {
            Timeout = TimeSpan.FromSeconds(timeoutSeconds),
        };
    }

    public void Dispose() => _http.Dispose();

    /// <summary>Embed a single string. Returns a float array of length ``dim``.</summary>
    public float[] Embed(string text)
    {
        var batch = EmbedBatch(new[] { text });
        return batch[0];
    }

    /// <summary>Embed a batch of strings. Preserves input order. Validates shape + finite + non-zero norm.</summary>
    public float[][] EmbedBatch(IReadOnlyList<string> texts)
    {
        if (texts.Count == 0)
        {
            return Array.Empty<float[]>();
        }

        EmbedResponse? body;
        try
        {
            using var req = new HttpRequestMessage(HttpMethod.Post, $"{_baseUrl}/api/embed")
            {
                Content = JsonContent.Create(
                    new EmbedRequest(_model, texts.ToArray(), _keepAlive),
                    EmbeddingsJsonContext.Default.EmbedRequest),
            };
            using var resp = _http.Send(req, HttpCompletionOption.ResponseContentRead);
            resp.EnsureSuccessStatusCode();
            using var stream = resp.Content.ReadAsStream();
            body = JsonSerializer.Deserialize(stream, EmbeddingsJsonContext.Default.EmbedResponse);
        }
        catch (HttpRequestException ex)
        {
            throw new EmbeddingException($"Ollama request failed: {ex.Message}", ex);
        }
        catch (TaskCanceledException ex)
        {
            throw new EmbeddingException($"Ollama request timed out: {ex.Message}", ex);
        }
        catch (JsonException ex)
        {
            throw new EmbeddingException($"Ollama returned non-JSON: {ex.Message}", ex);
        }

        if (body?.Embeddings is null || body.Embeddings.Length != texts.Count)
        {
            throw new EmbeddingException(
                $"Ollama returned {(body?.Embeddings?.Length.ToString() ?? "?")} embeddings for {texts.Count} inputs");
        }

        var matrix = new float[body.Embeddings.Length][];
        for (int i = 0; i < body.Embeddings.Length; i++)
        {
            var row = body.Embeddings[i];
            if (row is null)
            {
                throw new EmbeddingException("Ollama returned null embedding row");
            }
            double sumSq = 0;
            for (int j = 0; j < row.Length; j++)
            {
                var v = row[j];
                if (!float.IsFinite(v))
                {
                    throw new EmbeddingException("Ollama returned non-finite values");
                }
                sumSq += v * v;
            }
            if (sumSq == 0.0)
            {
                throw new EmbeddingException("Ollama returned zero-norm vector");
            }
            matrix[i] = row;
        }
        return matrix;
    }

    /// <summary>Diagnose the full Ollama path without throwing. Mirrors Python probe().</summary>
    public ProbeResult Probe()
    {
        string? version = null;
        try
        {
            using var resp = _http.Send(
                new HttpRequestMessage(HttpMethod.Get, $"{_baseUrl}/api/version"));
            resp.EnsureSuccessStatusCode();
            using var s = resp.Content.ReadAsStream();
            var v = JsonSerializer.Deserialize(s, EmbeddingsJsonContext.Default.VersionResponse);
            version = v?.Version;
        }
        catch (Exception ex) when (ex is HttpRequestException or TaskCanceledException or JsonException)
        {
            return new ProbeResult(false, null, false, false, null,
                $"cannot reach Ollama at {_baseUrl}: {ex.Message}");
        }

        try
        {
            using var resp = _http.Send(
                new HttpRequestMessage(HttpMethod.Get, $"{_baseUrl}/api/tags"));
            resp.EnsureSuccessStatusCode();
            using var s = resp.Content.ReadAsStream();
            var tags = JsonSerializer.Deserialize(s, EmbeddingsJsonContext.Default.TagsResponse);
            var names = tags?.Models?.Select(m => (m.Name ?? string.Empty).Split(':')[0]).ToHashSet(
                StringComparer.OrdinalIgnoreCase) ?? new HashSet<string>();
            var modelBase = _model.Split(':')[0];
            if (!names.Contains(modelBase))
            {
                return new ProbeResult(true, version, false, false, null,
                    $"model '{_model}' not pulled. Run: ollama pull {_model}");
            }
        }
        catch (Exception ex) when (ex is HttpRequestException or TaskCanceledException or JsonException)
        {
            return new ProbeResult(true, version, false, false, null,
                $"cannot list Ollama models: {ex.Message}");
        }

        try
        {
            var vec = Embed("probe");
            return new ProbeResult(true, version, true, true, vec.Length, null);
        }
        catch (EmbeddingException ex)
        {
            return new ProbeResult(true, version, true, false, null,
                $"embed probe failed: {ex.Message}");
        }
    }
}

// ---- BLOB codec + cosine --------------------------------------------------

internal static class Vectors
{
    /// <summary>Serialize a float vector as little-endian float32 bytes.</summary>
    public static byte[] Pack(float[] v)
    {
        var bytes = new byte[v.Length * sizeof(float)];
        for (int i = 0; i < v.Length; i++)
        {
            BinaryPrimitives.WriteSingleLittleEndian(
                bytes.AsSpan(i * sizeof(float), sizeof(float)), v[i]);
        }
        return bytes;
    }

    /// <summary>Deserialize a BLOB produced by Pack back to a float array.</summary>
    public static float[] Unpack(ReadOnlySpan<byte> blob, int dim)
    {
        if (blob.Length != dim * sizeof(float))
        {
            throw new ArgumentException(
                $"blob length {blob.Length} does not match expected {dim * sizeof(float)} for dim={dim}");
        }
        var v = new float[dim];
        for (int i = 0; i < dim; i++)
        {
            v[i] = BinaryPrimitives.ReadSingleLittleEndian(
                blob.Slice(i * sizeof(float), sizeof(float)));
        }
        return v;
    }

    /// <summary>
    /// Cosine similarity between a query vector and each row of the candidate
    /// matrix. Uses Vector&lt;float&gt; SIMD where available.
    /// </summary>
    public static float[] CosineMatrix(float[] query, IReadOnlyList<float[]> candidates)
    {
        if (query.Length == 0 || candidates.Count == 0)
        {
            return new float[candidates.Count];
        }
        var qNorm = MathF.Sqrt(DotSelf(query));
        var result = new float[candidates.Count];
        if (qNorm == 0f)
        {
            return result;
        }
        for (int i = 0; i < candidates.Count; i++)
        {
            var c = candidates[i];
            if (c.Length != query.Length)
            {
                result[i] = 0f;
                continue;
            }
            var cNorm = MathF.Sqrt(DotSelf(c));
            if (cNorm == 0f)
            {
                result[i] = 0f;
                continue;
            }
            var dot = Dot(query, c);
            result[i] = dot / (qNorm * cNorm);
        }
        return result;
    }

    private static float Dot(float[] a, float[] b)
    {
        int n = a.Length;
        int simdWidth = Vector<float>.Count;
        float sum = 0f;
        int i = 0;
        if (Vector.IsHardwareAccelerated && n >= simdWidth)
        {
            var acc = Vector<float>.Zero;
            for (; i <= n - simdWidth; i += simdWidth)
            {
                var va = new Vector<float>(a, i);
                var vb = new Vector<float>(b, i);
                acc += va * vb;
            }
            sum = Vector.Sum(acc);
        }
        for (; i < n; i++)
        {
            sum += a[i] * b[i];
        }
        return sum;
    }

    private static float DotSelf(float[] a)
    {
        int n = a.Length;
        int simdWidth = Vector<float>.Count;
        float sum = 0f;
        int i = 0;
        if (Vector.IsHardwareAccelerated && n >= simdWidth)
        {
            var acc = Vector<float>.Zero;
            for (; i <= n - simdWidth; i += simdWidth)
            {
                var va = new Vector<float>(a, i);
                acc += va * va;
            }
            sum = Vector.Sum(acc);
        }
        for (; i < n; i++)
        {
            sum += a[i] * a[i];
        }
        return sum;
    }
}

// ---- JSON contracts (source-generated for AOT) ----------------------------

internal sealed class EmbedRequest
{
    [JsonPropertyName("model")] public string Model { get; init; }
    [JsonPropertyName("input")] public string[] Input { get; init; }
    // Issue #11: null values get omitted by System.Text.Json defaults
    // (DefaultIgnoreCondition = WhenWritingNull on the context below), so
    // keep_alive is only sent when the client was configured with one.
    [JsonPropertyName("keep_alive")] public string? KeepAlive { get; init; }

    public EmbedRequest(string model, string[] input, string? keepAlive = null)
    {
        Model = model;
        Input = input;
        KeepAlive = keepAlive;
    }
}

internal sealed class EmbedResponse
{
    [JsonPropertyName("embeddings")] public float[]?[]? Embeddings { get; set; }
}

internal sealed class VersionResponse
{
    [JsonPropertyName("version")] public string? Version { get; set; }
}

internal sealed class TagsResponse
{
    [JsonPropertyName("models")] public TagModel[]? Models { get; set; }
}

internal sealed class TagModel
{
    [JsonPropertyName("name")] public string? Name { get; set; }
}

[JsonSourceGenerationOptions(
    // Drop keep_alive from the serialized JSON when null, so the existing
    // contract (no keep_alive field = Ollama default 5m) is preserved for
    // clients that don't set one. Issue #11.
    DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull)]
[JsonSerializable(typeof(EmbedRequest))]
[JsonSerializable(typeof(EmbedResponse))]
[JsonSerializable(typeof(VersionResponse))]
[JsonSerializable(typeof(TagsResponse))]
[JsonSerializable(typeof(TagModel))]
internal partial class EmbeddingsJsonContext : JsonSerializerContext
{
}
